### PR TITLE
Store vector layer wkb type in xml

### PR DESCRIPTION
**Pull request by nyalldawson**
We can use this when restoring the layer, if the uri turns out
to be invalid at that stage (e.g. a file has moved). By storing
and falling back to the last known wkb type, we avoid unnecessarily
discarding the existing layer renderer, and can still show the
expected layer type in the layer tree.

### CODE REVIEW

**Comment by nyalldawson** [comment]
@elpaso also as discussed - turns out we DON'T yet store the last known wkb type.

File: src/core/qgsvectorlayer.cpp [hunk 1]
```diff
@@ -1705,6 +1710,7 @@ bool QgsVectorLayer::writeXml( QDomNode &layer_node,
 
   // set the geometry type
   mapLayerNode.setAttribute( QStringLiteral( "geometry" ), QgsWkbTypes::geometryDisplayString( geometryType() ) );
+  mapLayerNode.setAttribute( QStringLiteral( "wkbType" ), wkbType() );
```
Comment (m-kuhn) [inline src/core/qgsvectorlayer.cpp]:
```suggestion
  mapLayerNode.setAttribute( QStringLiteral( "wkbType" ), qgsEnumValueToKey( wkbType() ) );
```

File: src/core/qgsvectorlayer.cpp [hunk 2]
```diff
@@ -1418,6 +1418,11 @@ bool QgsVectorLayer::readXml( const QDomNode &layer_node, QgsReadWriteContext &c
   if ( !setDataProvider( mProviderKey, options ) )
   {
     QgsDebugMsg( QStringLiteral( "Could not set data provider for layer %1" ).arg( publicSource() ) );
+    const QDomElement elem = layer_node.toElement();
+
+    // for invalid layer sources, we fallback to stored wkbType if available
+    if ( elem.hasAttribute( QStringLiteral( "wkbType" ) ) )
+      mWkbType = static_cast< QgsWkbTypes::Type >( elem.attribute( QStringLiteral( "wkbType" ) ).toInt() );
```
Comment (m-kuhn) [inline src/core/qgsvectorlayer.cpp]:
```suggestion
      mWkbType = qgsEnumKeyToValue( elem.attribute( QStringLiteral( "wkbType" ) ).toString() );
```

**Review by m-kuhn — APPROVED** [review]
Looks good :+1:

**Review by elpaso — APPROVED** [review]
Thanks @nyalldawson !

**Pull request merged**